File: src/main/go/leetcode/lists/offer/Offer22.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2022/12/28 23:41
 */

func getKthFromEnd(head *ListNode, k int) *ListNode {
	dummy := &ListNode{}
	dummy.Next = head
	pre := dummy
	length := 0
	cur := head
	for cur != nil {
		length++
		cur = cur.Next
	}

	step := length - k
	for step > 0 {
		pre = pre.Next
		step--
	}
	return pre.Next
}

File: java/src/leetcode/lists/offer/Offer22.java
package leetcode.lists.offer;

import common.ListNode;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/12 21:39
 */
public class Offer22 {

    public ListNode getKthFromEnd(ListNode head, int k) {
        if (head == null) {
            return null;
        }
        ListNode fast = head;
        while (k-- > 0) {
            if (fast == null) {
                return null;
            }
            fast = fast.next;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode cur = dummy;
        while (fast != null) {
            cur = cur.next;
            fast = fast.next;
        }
        return cur.next;
    }
}
